File: src/graphite.cpp
#include <iostream>

#include "../include/graphite.h"
#include "../include/ImputeByRef.h"
#include "../include/ImputeNoRef.h"
#include "../include/materials.h"
#include "../include/option.h"
#include "../include/error_codes.h"


//////////////////// graphite ////////////////////

void graphite(const Option& option) {
	option.print_info();
	Materials	*materials = Materials::create(option.path_map,
													option.path_ped);
	materials->display_map_info();
	
	VCFHuge	*vcf = NULL;
	try {
		vcf = VCFHuge::read(option.path_vcf);
		if(option.exists_ref())
			ImputeByRef::impute(vcf, materials, option);
		else
			ImputeNoRef::impute(vcf, materials, option);
		delete vcf;
	}
	catch(const ExceptionWithCode& e) {
		if(vcf != NULL)
			delete vcf;
		delete materials;
		throw;
	}
	delete materials;
}

int main(int argc, char **argv) {
	const Option	*option = Option::create(argc, argv);
	if(option == NULL) {
		Option::usage(argv);
		exit(ErrorCode::WRONG_ARGUMENT);
	}
	
	try {
		graphite(*option);
	}
	catch(const ExceptionWithCode& e) {
		delete option;
		std::cerr << e.what() << std::endl;
		exit(e.get_error_code());
	}
	
	delete option;
	return 0;
}

File: python/graphite.py
from __future__ import annotations

# coding: utf-8
# graphite.py

from itertools import *
import sys

import ImputeByRef
import ImputeNoRef
from VCF import VCFHuge
from materials import *
from Map import *
from option import *
from exception_with_code import *


#################### process ####################

def graphite(option: Option) -> None:
	option.print_info()
	materials = Materials.create(option.path_map, option.path_ped)
	materials.display_map_info()
	vcf = VCFHuge.read(option.path_VCF)
	
	if option.exists_ref():
		ImputeByRef.impute(vcf, materials, option)
	else:
		ImputeNoRef.impute(vcf, materials, option)


#################### main ####################

option = Option.create(sys.argv)
try:
	if option is None:
		Option.usage()
		exit(1)
	else:
		graphite(option)
		exit(0)
except ExceptionWithCode as e:
	print(str(e), file=sys.stderr)
	exit(e.get_error_code().value)
